### PR TITLE
added: (Tensor|Vec)Func::timeDerivative 

### PR DESCRIPTION
**Pull request by akva2**
returns the time derivative of the function

Sits on top of https://github.com/OPM/IFEM/pull/641

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 ifem-elasticity=150 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 ifem-elasticity=150 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 ifem-elasticity=150 please

File: src/Utility/Function.h [hunk 1]
```diff
@@ -262,6 +269,19 @@ class VecFunc : public utl::SpatialFunction<Vec3>, public FunctionBase
     result = this->evalGradient(X);
     return result;
   }
+
+  //! \brief Evaluates time derivatives of the function.
+  Vec3 timeDerivative(const Vec3& X) const
+  {
+    return this->evalTimeDerivative(X);
+  }
```
Comment (kmokstad) [inline src/Utility/Function.h:277]: Define on single line

**Review by kmokstad — APPROVED** [review]
Looks OK, as far I can see, although it starts to be quite a complex Function hierarchy with many evaluation methods for derivatives on different forms, so it is hard to get the whole picture in a single commit snapshot only showing the differences.

File: src/Utility/ExprFunctions.C [hunk 2]
```diff
@@ -499,6 +499,18 @@ EvalMultiFunction<ParentFunc, Ret>::evalGradient (const Vec3& X) const
 }
 
 
+template <class ParentFunc, class Ret>
+std::vector<Real>
+EvalMultiFunction<ParentFunc, Ret>::evalTimeDerivative (const Vec3& X) const
+{
+  std::vector<Real> result(this->ncmp);
+  for (size_t i = 1; i <= this->ncmp; ++i)
+    result[i-1] = this->p[i-1]->deriv(X,4);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/ExprFunctions.C]: ah yes, ranged for it is.

File: src/Utility/TensorFunction.C [hunk 3]
```diff
@@ -23,6 +23,15 @@ utl::matrix3d<Real> TensorFunc::gradient(const Vec3& X) const
 }
 
 
+Tensor TensorFunc::timeDerivative(const Vec3& X) const
+{
+  const size_t nsd = sqrt(this->dim());
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/TensorFunction.C]: done

File: src/Utility/Function.h [hunk 4]
```diff
@@ -262,6 +269,19 @@ class VecFunc : public utl::SpatialFunction<Vec3>, public FunctionBase
...
+
+protected:
+  //! \brief Returns the time derivatives of the function as a 1D array.
+  virtual std::vector<Real> evalTimeDerivative(const Vec3& X) const
+  {
+    return std::vector<Real>(ncmp);
+  }
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/Function.h]: The values would be zero initialized, but in any case; I didn't check if the the Vec3 constructor handles empties and did the defensive but it does so yeah, dropped the dummy.

File: src/Utility/Function.h (hunk 1 above)
Comment (akva2) [reply src/Utility/Function.h]: single-lined although it gets to 82 chars.

File: src/Utility/Function.h [hunk 5]
```diff
@@ -220,6 +222,11 @@ class RealFunc : public utl::SpatialFunction<Real>, public FunctionBase
 
   //! \brief Returns a representative scalar equivalent of the function value.
   virtual Real getScalarValue(const Vec3& X) const { return this->evaluate(X); }
+
+  Real timeDerivative(const Vec3& X) const
+  {
+    return this->deriv(X,4);
+  }
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/Function.h]: doxy has been added and method single-lined.

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-elasticity=150 please

File: src/Utility/Function.h (hunk 1 above)
Comment (kmokstad) [reply src/Utility/Function.h]: yep, that's the price of using a bit long'ish method names

**Pull request merged**